cobra/util/solver.py

```diff
@@ -83,44 +83,52 @@ def set_objective(model, value, additive=False):
         If true, add the terms to the current objective, otherwise start with
         an empty objective.
     """
-    by_objective = isinstance(value,
-                              (sympy.Basic, model.solver.interface.Objective))
-
-    not_supported = (
-        additive and (not model.objective.is_Linear and not by_objective))
-    if not_supported:
-        raise ValueError('can only update non-linear objectives additively '
-                         'using object of class '
-                         'model.solver.interface.Objective, not %s' %
-                         type(value))
-    reverse_value = None
-    if by_objective:
-        if not additive:
-            if isinstance(value, sympy.Basic):
-                value = model.solver.interface.Objective(value, sloppy=False)
-            model.solver.objective = value
-        else:
-            if isinstance(value, model.solver.interface.Objective):
-                value = value.expression
-            model.solver.objective += value
-            reverse_value = -value
-    elif isinstance(value, dict):
+    if isinstance(value, dict):
+        if not model.objective.is_Linear:
+            raise ValueError('can only update non-linear objectives '
+                             'additively using object of class '
+                             'model.solver.interface.Objective, not %s' %
+                             type(value))
         if not additive:
             model.solver.objective = model.solver.interface.Objective(
                 sympy.S.Zero, direction='max')
```

Comment:
Wouldn't it be better if that would respect the direction from the previous objective?